src/Generator/SequentialInvoiceNumberGenerator.php

```diff
@@ -49,15 +49,22 @@ public function __construct(
 
     public function generate(): string
     {
-        $invoiceIdentifierPrefix = $this->dateTimeProvider->__invoke()->format('Y/m') . '/';
+        $currentDate = $this->dateTimeProvider->__invoke();
+
+        $invoiceIdentifierPrefix = $currentDate->format('Y/m') . '/';
 
         /** @var InvoiceSequenceInterface $sequence */
         $sequence = $this->getSequence();
 
         $this->sequenceManager->lock($sequence, LockMode::OPTIMISTIC, $sequence->getVersion());
 
+        if ($this->isMonthChanged($currentDate, $sequence->getLastGeneratedAt())) {
+            $sequence->resetIndex();
```

Comment:
This might cause issues when events are processed asynchronously and not in chronological order, eg. you've got OrderPlaced at 31.03, OrderPlaced at 01.04 and OrderPlaced at 31.03 again.

The solution I'd see for that is to replace `lastGeneratedAt` with `group` (which will be `Y/m`) and always get the sequence for a given group. It removes the need for this conditional and resetting index also.